openrtb-json/src/main/scala/com/powerspace/openrtb/json/BidRequestSerde.scala

```diff
@@ -1,17 +1,114 @@
+import com.google.openrtb.BidRequest.{Imp, Source}
+import com.powerspace.openrtb.json.util.JsonUtils
+import io.circe.generic.extras.Configuration
+
+
 
 /**
   * Serialize and Deserialize an OpenRTB BidRequest
   */
 object BidRequestSerde {
+  import io.circe._
+  import io.circe.syntax._
+  import io.circe.generic.extras.semiauto._
+
+  import JsonUtils._
+  implicit val customConfig: Configuration = Configuration.default.withSnakeCaseMemberNames
+
+  implicit val bannerAdTypeEncoder: Encoder[BannerAdType] = protobufEnumEncoder[BannerAdType]
+  implicit val creativeAttributeEncoder: Encoder[CreativeAttribute] = protobufEnumEncoder[CreativeAttribute]
+  implicit val adPositionEncoder: Encoder[AdPosition] = protobufEnumEncoder[AdPosition]
+  implicit val expandableDirectionEncoder: Encoder[ExpandableDirection] = protobufEnumEncoder[ExpandableDirection]
+  implicit val apiFrameworkEncoder: Encoder[APIFramework] = protobufEnumEncoder[APIFramework]
+
+  implicit val protocolEncoder: Encoder[Protocol] = protobufEnumEncoder[Protocol]
+  implicit val videoPlacementTypeEncoder: Encoder[VideoPlacementType] = protobufEnumEncoder[VideoPlacementType]
+  implicit val videoLinearityEncoder: Encoder[VideoLinearity] = protobufEnumEncoder[VideoLinearity]
+  implicit val playbackMethodEncoder: Encoder[PlaybackMethod] = protobufEnumEncoder[PlaybackMethod]
+  implicit val playbackCessationModeEncoder: Encoder[PlaybackCessationMode] = protobufEnumEncoder[PlaybackCessationMode]
+  implicit val contentDeliveryMethodEncoder: Encoder[ContentDeliveryMethod] = protobufEnumEncoder[ContentDeliveryMethod]
+  implicit val companionTypeEncoder: Encoder[CompanionType] = protobufEnumEncoder[CompanionType]
+
+  implicit val feedTypeEncoder: Encoder[FeedType] = protobufEnumEncoder[FeedType]
+  implicit val volumeNormalizationModeEncoder: Encoder[VolumeNormalizationMode] = protobufEnumEncoder[VolumeNormalizationMode]
+  implicit val auctionTypeEncoder: Encoder[AuctionType] = protobufEnumEncoder[AuctionType]
+
+  implicit val companionAdEncoder: Encoder[CompanionAd] = deriveEncoder[CompanionAd].clean
+  implicit val formatEncoder: Encoder[Format] = deriveEncoder[Format].clean
+  implicit val bannerEncoder: Encoder[Banner] = deriveEncoder[Banner].recodeBooleans.clean
+  implicit val videoEncoder: Encoder[Video] =deriveEncoder[Video].recodeBooleans.clean
+  implicit val sourceEncoder: Encoder[Source] = deriveEncoder[Source].clean()
+  implicit val audioEncoder: Encoder[Audio] = deriveEncoder[Audio].recodeBooleans.clean
+  implicit val pmpEncoder: Encoder[Pmp] = deriveEncoder[Pmp].recodeBooleans.clean
+  implicit val dealEncoder: Encoder[Deal] = deriveEncoder[Deal].recodeBooleans.clean
+
+  /**
+    * Encoder for the OpenRTB native object.
+    */
+  implicit val nativeEncoder: Encoder[Native] =
+    NoNullsEncoder(native => Seq(
+      //("request", native.requestOneof.value.asJson), //@todo: to be verified
```

Comment:
You may have to provide the case when request is a string.       

```
message Native {
      oneof request_oneof {
        // Request payload complying with the Native Ad Specification.
        // Exactly one of {request, request_native} should be used;
        // this is the OpenRTB-compliant field for JSON serialization.
        string request = 1;

        // Request payload complying with the Native Ad Specification.
        // Exactly one of {request, request_native} should be used;
        // this is an alternate field preferred for Protobuf serialization.
        NativeRequest request_native = 50;
      }
``` 

I think you can use a customized encoder and that the rest of encoder can be derived.